common/build/eslint-plugin-fluid/src/rules/no-unchecked-record-access.js

```diff
@@ -0,0 +1,358 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/**
+ * Rule to enforce safe property access on index signature types.
+ *
+ * Reports issues when [non-array] index properties are accessed without handling
+ * the possibility that they are absent.
+ * Enabling `noUncheckedIndexedAccess` will disable these checks.
+ */
+
+const { SyntaxKind } = require("typescript");
+
+module.exports = {
+	meta: {
+		type: "problem",
+		docs: {
+			description: "Disallow unchecked property access on index signature types",
+			category: "Possible Errors",
+		},
+		schema: [],
+	},
+	create(context) {
+		const parserServices = context.parserServices;
+
+		// Check if we have the necessary TypeScript services
+		if (!parserServices || !parserServices.program || !parserServices.esTreeNodeToTSNodeMap) {
+			return {};
+		}
+		const compilerOptions = parserServices.program.getCompilerOptions();
+		const typeChecker = parserServices.program.getTypeChecker();
+
+		// If noUncheckedIndexedAccess is already enabled, disable this rule
+		if (compilerOptions.noUncheckedIndexedAccess) {
+			return {};
+		}
+
+		// Main function to run on every member access (e.g., obj.a or obj["a"])
+		function checkPropertyAccess(node) {
+			// Only check index signature types
+			if (!isIndexSignatureType(parserServices, node)) {
+				return;
+			}
+
+			// Skip if the property has been checked (e.g., with optional chaining), skip it. Please see isDefined() for exhaustive list.
+			if (propertyHasBeenChecked(node)) {
+				return;
+			}
+
+			const fullName = getFullName(node);
+			const parentNode = node.parent;
+
+			/*
+			 * Cases when this lint rule should report an error
+			 */
```

Comment:
nit: "error" is used a lot in here. Technically these are "defects". The related configuration will determine if a defect is an error or a warning (or ignored).